cpp/src/bitmask/bit_mask.h

```diff
@@ -0,0 +1,291 @@
+      return (record_idx % BITS_PER_ELEMENT);
+    }
+  }
+
+  /**
+   * @brief determine the number of bit_mask_t elements are used
+   *
+   * @param[in]  size    Number of bits in the bitmask
+   *
+   * @return the number of elements
+   */
+  CUDA_HOST_DEVICE_CALLABLE
+  constexpr gdf_size_type num_elements(gdf_size_type size) { 
+    return (( size + ( details::BITS_PER_ELEMENT - 1)) / details::BITS_PER_ELEMENT ); 
+  }
+
+  /**
+   *  @brief Copy data between host and device
+   *
+   *  @param[out] dst      - the address of the destination
+   *  @param[in] src       - the address of the source
+   *  @param[in] num_bits  - the number of bits in the bit container
+   *  @param[in] kind      - the direction of the copy
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error copy_bit_mask(bit_mask_t *dst, const bit_mask_t *src, size_t num_bits, enum cudaMemcpyKind kind) {
+    CUDA_TRY(cudaMemcpy(dst, src, num_elements(num_bits) * sizeof(bit_mask_t), kind));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   *  @brief Deallocate device space for the valid bit mask
+   *
+   *  @param[in]  valid   The pointer to device space that we wish to deallocate
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error destroy_bit_mask(bit_mask_t *valid) {
+    RMM_TRY(RMM_FREE(valid, 0));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   *  @brief Get a single element of bits from the device
+   *
+   *  @param[out]  element - address on host where the bits will be stored
+   *  @param[out]  device_element - address on the device containing the bits to fetch
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error get_element(bit_mask_t *element, const bit_mask_t *device_element) {
+    CUDA_TRY(cudaMemcpy(element, device_element, sizeof(bit_mask_t), cudaMemcpyDeviceToHost));
+    return GDF_SUCCESS;
+  }
+	  
+  /**
+   *  @brief Put a single element of bits to the device
+   *
+   *  @param[out]  element - address on host containing bits to store
+   *  @param[out]  device_element - address on the device where the bits will be stored
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error put_element(bit_mask_t element, bit_mask_t *device_element) {
+    CUDA_TRY(cudaMemcpy(device_element, &element, sizeof(bit_mask_t), cudaMemcpyHostToDevice));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   *  @brief Allocate device space for the valid bit mask.
+   *
+   *  @param[out] mask                  address of the bit mask pointer
+   *  @param[in]  number_of_records     number of records
+   *  @param[in]  fill_value            optional, should the memory be initialized to all 0 or 1s. All other
+   *                                    values indicate un-initialized.  Default is uninitialized
+   *  @param[in]  padding_bytes         optional, specifies byte boundary the data should be padded to.
```

Comment:
Do we have a use case for `padding_bytes` currently? I want to make sure this isn't a feature without a user. 
cudaMalloc already (internally) aligns all allocations to a fairly large granularity (like 512 bytes or something). RMM does this too (via cnmem).

Reply:
Jake indicated this was a requirement for Arrow.  I don't know anything more about it.

Reply:
@harrism the Arrow spec requires 64 byte padding. If you rely on the cudaMalloc 512B alignment, you'll still get cuda-memcheck warnings (as you've already seen).

Reply:
OK, so is there code that uses this? Or is it unused/untested?

Reply:
I don't have a test that specifies a value for padding_bytes other than the default.

Reply:
@harrism I'm not sure what you mean by code that "uses" the padding. It's just part of the requirements.

Reply:
If there are requirements, then we must plan to call the function with something other than the default padding. Otherwise it's a feature with no user. At the least, please add a unit test that exercises this feature.

Reply:
Ah, I see, you're referring to it being a parameter rather than just being the default behavior (not a parameter). I'd be fine with removing it as a parameter and just hardcoding the padding at 64B.